### ISSUE TITLE
scalar categoricals are sometimes interpreted as data keys

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#9844]
### Bug report

**Bug summary**

**Code for reproduction**

```python
from matplotlib import pyplot as plt
fig, axs = plt.subplots(2)
axs[0].bar("thing", 1, data={"thing": 1})
axs[1].bar("other", 1, data={"thing": 1})
plt.show()
```

**Actual outcome**

![figure_1](https://user-images.githubusercontent.com/1322974/33191136-cb8e3f06-d06a-11e7-9f13-6b70df269831.png)

The first plot's x-value ("thing") is interpreted as a lookup into the data dict and replaced by 1.
The second plot's x-value does not appear in the data dict and is interpreted as a categorical.

**Expected outcome**

Not sure what the best option is, but silently changing the interpretation of input based on whether it is present or not in another dict seems a bit finnicky.  I have proposed in other places to not allow scalar categoricals at all (always need to be passed in a container -- list, array, dataframe, etc.), which also solves the inconsistency that `plot(1, "x")` currently specs a marker whereas `plot("x", 1)` treats `"x"` as a categorical.

Unlike other categorical issues I don't actually think this is release critical per se, but it would still be nice to get the behavior clarified/simplified...

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 2.1
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version:
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by story645** [comment]
This behavior happens with any scalar, and so I'm wondering if the solution isn't something like locking out scaler plotting that aren't keys in data if data is present. 
    
```python
from matplotlib import pyplot as plt
fig, axs = plt.subplots(2)
axs[0].bar(10, 1, data={10: 1})
axs[1].bar(20, 1, data={10: 1})
plt.show()
```
![index](https://user-images.githubusercontent.com/1300499/33334156-03a1f2c8-d437-11e7-9243-7d75c4204508.png)

**Comment by story645** [comment]
And looking the example I put down, it's even more buggy 'cause it's not registering the scaler as the key at all (floating point error?).

**Comment by ksunden** [comment]
I'm not sure there is much that _can_ be done here... 

There are two rules that come into play with handling strings as data:

- The data kwarg handling, which uses the string to access the value of a mapping
- The categorical handling

Necessarily, they must be resolved in that order. If they were resolved in the other order, categorical handling would _always_ take place, and thus the data kwarg would be useless.

As for non-strings as keys of the data kwarg, I think that is unsupported undefined behavior:

The docstring entry for `data` kwarg reads:

> dataindexable object, optional
>
>    If given, all parameters also accept a string s, which is interpreted as data[s] (unless this raises an exception).

As of (at least) [#10928](https://github.com/matplotlib/matplotlib/pull/10928), that the keys are strings to have any data kwarg behavior is strictly enforced

I suppose the "not allowed to pass categoricals as scalars" is potentially a solution (as that would mean that strings are not expected for any parameter except via the data kwarg)... but that is a rather large API change for a rather narrow set of interactions, though perhaps it would have other benefits (and I'm not quite sure how easy it would be to change that, actually).

If we are unwilling to deprecate using bare strings for either of those cases, though, I don't see any other behavior change, perhaps some docs changes, but even then not sure where.

**Comment by timhoffm** [comment]
IMHO this is a lost cause. There are cases outside of categoricals, in which data-replaceable args can rightfully be strings. Consider the following realistic example
```
from matplotlib import pyplot as plt
fig, axs = plt.subplots(2)
data = {
    "x": [0, 1],
    "y": [1, 1], 
    "color": ["c", "m"],
}
axs[0].scatter("x", "y", facecolor="color", data=data)
axs[1].scatter("x", "y", facecolor="red", data=data)
plt.show()
```
![grafik](https://github.com/matplotlib/matplotlib/assets/2836374/5d1cea4c-11c1-4ac2-9f05-1ebc3697705e)

We cannot reasonably prohibit the second case. As a consequence, it would be very tedious to tell acceptable string scalars from illegal string scalars apart.

I propose to close this. And while "silently changing the interpretation of input based on whether it is present or not in another dict is a bit finnicky" is, the behavior is consistent: "If a string, try to look it up in ``data`` otherwise proceed as normal". A mild improvement could be updating ``data`` description from

> If given, the following parameters also accept a string `s`, which is interpreted as `data[s]` (unless this raises an exception):

to

> If given, the following parameters also accept a string `s`, which is interpreted as `data[s]` if `data[s]` exists.

**Comment by anntzer** [comment]
Indeed, this is a good example regarding the difficulty to fix this.
I'm not sure about the proposed wording change; what does "`data[s]` exists" actually mean? (other than "the expression does not raise an exception"...)

**Comment by timhoffm** [comment]
You're right it's not much clearer. I find the parentheses in the original message a bit confusing, because (1) why put in parentheses (2) what happens if that raises?  What we need to communicate, is

> If the parameter is a string `s`, we try to look up `data[s]`. If that works, the resulting value is used for the parameter. If it fails, `s` itself is used (which may or may not be a valid input type for the parameter).

Can't think of a nice wording for this right now. Help welcome.

**Comment by jklymak** [comment]
Is it not just "if _s_ is a key of _data_"?

**Comment by timhoffm** [comment]
I'm not quite sure that *key* is universally understood. For example I don't think columns of a DataFrame or fields of a structured numpy array are commonly referred to as *keys*.

**Comment by jklymak** [comment]
Certainly pandas and xarray both refer to _keys_, and that is the terminology for dictionaries.  If people don't know that term when using data structures they should probably learn. However I agree it is also helpful to have data[s].  My phrasing above was not meant to replace the whole sentence, but is less awkward than "data[s] raises and exception".
